server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
-                                                                    );
-                                                                }
-                                                                // acknowledged maybe be false but some indices may have been correctly
-                                                                // closed, so we maintain a kind of coherency by overriding the
-                                                                // shardsAcknowledged value (see ShardsAcknowledgedResponse constructor)
-                                                                boolean shardsAcked = acknowledged ? shardsAcknowledged : false;
-                                                                listener.onResponse(
-                                                                    new CloseIndexResponse(acknowledged, shardsAcked, indices)
-                                                                );
-                                                            },
-                                                            listener::onFailure
-                                                        );
-                                                    } else {
-                                                        listener.onResponse(new CloseIndexResponse(acknowledged, false, indices));
-                                                    }
-                                                }
-                                            },
-                                            ClusterStateTaskExecutor.unbatched()
-                                        ),
-                                        listener::onFailure
-                                    )
-                                )
-                            );
-                    }
-                }
+    private class CloseIndicesExecutor implements ClusterStateTaskExecutor<CloseIndicesTask> {
 
-                @Override
-                public void onFailure(final Exception e) {
-                    listener.onFailure(e);
+        @Override
+        public ClusterTasksResult<CloseIndicesTask> execute(ClusterState currentState, List<CloseIndicesTask> tasks) throws Exception {
+            ClusterTasksResult.Builder<CloseIndicesTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            for (CloseIndicesTask task : tasks) {
+                try {
+                    Map<Index, ClusterBlock> blockedIndices = task.blockedIndices;
+                    Map<Index, IndexResult> verifyResults = task.verifyResults;
+                    final List<IndexResult> indices = new ArrayList<>();
+
+                    Tuple<ClusterState, Collection<IndexResult>> closingResult = closeRoutingTable(state, blockedIndices, verifyResults);
+                    assert verifyResults.size() == closingResult.v2().size();
+                    indices.addAll(closingResult.v2());
```

Comment:
Do we need to copy `closingResult.v2()` to a new list here? Think we could just use it directly in the publish listener.

Reply:
d3445d2a27bad2477845c81d794c205a80f00595 and then d64a1037c7dac962bc309ea49fed212ffa0e7854